annotate.el

```diff
@@ -1073,6 +1090,116 @@ example:
         (when annotate-use-messages
           (message "Annotations loaded."))))))
 
+(defun annotate-db-clean-records (records-db)
+  "Remove records from arg `records-db' that have empty annotation, example:
+
+'((\"/foo/bar.dat\" nil \"abababababababababababababab\")
+  (\"/foo/baz.dat\" ((0 9 \"note\" \"annotated\")) \"abababababababababababababab\"))
+
+will become:
+
+'((\"/foo/baz.dat\" ((0 9 \"note\" \"annotated\")) \"abababababababababababababab\"))
+
+i.e. the first record is removed."
+  (cl-remove-if (lambda (a) (null (annotate-annotations-from-dump a)))
+                records-db))
+
+(defun annotate-db-purge ()
+ "Update datbase *on disk* removing all the records with empty
+annotation."
+  (interactive)
+  (let ((db (annotate-db-clean-records (annotate-load-annotation-data))))
+    (annotate-dump-annotation-data db)))
+
+(defun annotate-load-annotation-data ()
+  "Read and return saved annotations."
+  (with-temp-buffer
+    (when (file-exists-p annotate-file)
+      (insert-file-contents annotate-file))
+    (goto-char (point-max))
+    (cond ((= (point) 1)
+           nil)
+          (t
+           (goto-char (point-min))
+           (read (current-buffer))))))
+
+(defun annotate-dump-annotation-data (data)
+  "Save `data` into annotation file."
+  (with-temp-file annotate-file
+    (let ((print-length nil))
+      (prin1 data (current-buffer)))))
+
+(defun annotate-db-remove-annotation (db-records
+                                      record-filename
+                                      annotation-beginning
+                                      annotation-ending)
+  "Remove from database `db-records' the annotation indentified by
+ the triplets `record-filename', `annotation-beginning' and
+ `annotation-ending'; if such annotation does exists."
+  (let* ((filename-match-p (lambda (record)
+                             (string= (annotate-filename-from-dump record)
+                                      record-filename)))
+         (file-matched-record (cl-find-if filename-match-p db-records)))
+    (if file-matched-record
+        (let* ((rest-of-db      (cl-remove-if filename-match-p db-records))
```

Comment:
I kind of dislike the use of `let*`. I often find myself using `let*` as a simple way of writing "procedural" Python code in lisp.

In this case, I would construct `rest-of-db` and `new-annotation` in a plain `let`, and inline `checksum` and `new-record` into `push`. But I honestly don't know if this is idiomatic or not.

Reply:
>I kind of dislike the use of let*. I often find myself using let* as a simple way of writing "procedural" >Python code in lisp.

Indeed this (as far as i know!) is what is used for! :) Anyway consider that this form below:

``` lisp
(let* ((a 1)
       (b 2))
  (- a b))
```
is just syntactic sugar for:

``` lisp
(let ((a 1))
  (let ((b 2))
    (- a b)))
```
that, in turn, is just syntactic sugar for:

``` lisp
((lambda (a)
   ((lambda (b) (- a b))
    2))
 1)
```
So this is actually very lispy, i suppose! :D

Reply:
Yeah, it's cool how lisp just decomposes into plain old lambdas. For some reason, I sometimes empathize with the machine, and try to nest as few parenthesis as possible; i.e. try to use `let` instead of `let*`, as in this case. But I guess that's actually silly :-)